src/analysis_and_optimization/Mem_pattern.ml

```diff
@@ -0,0 +1,602 @@
+ * (TODO: We can probably coerce this and the 
+ *  other query demotable function together)
+ * 
+ *)
+let query_demotable_funkinds_loop query_expr_loop in_loop kind exprs =
+  let ret_demotable_types =
+    Set.Poly.union_list (List.map ~f:query_all_expr_eigen_names exprs)
+  in
+  match kind with
+  | Fun_kind.StanLib (name, (_ : bool Fun_kind.suffix), _) -> (
+    match name with
+    | x
+      when Stan_math_signatures.is_reduce_sum_fn x
+           || Stan_math_signatures.is_variadic_ode_fn x ->
+        ret_demotable_types
+    | _ ->
+        let is_fun_support =
+          Stan_math_signatures.query_stan_math_mem_pattern_support name
+            (List.map ~f:find_args exprs)
+        in
+        if is_fun_support then
+          Set.Poly.union_list (List.map ~f:(query_expr_loop in_loop) exprs)
+        else ret_demotable_types )
+  | CompilerInternal (_ : Internal_fun.t) -> Set.Poly.empty
+  | UserDefined ((_ : string), (_ : bool Fun_kind.suffix)) ->
+      ret_demotable_types
+
+(**
+ * Find indices on Matrix and Vector types that perform single 
+ *  cell access. Returns true if it finds
+ * a vector, row vector, matrix, or matrix with single cell access
+ * as well as an array of any of the above that is accessing the 
+ * inner matrix types cell.
+ *)
+let rec is_uni_eigen_loop_indexing (ut : UnsizedType.t)
+    (index : 'a Expr.Fixed.t Index.t list) =
+  match (ut, index) with
+  | (UnsizedType.UArray t | UFun (_, ReturnType t, _, _)), index -> (
+    match List.tl index with
+    | Some cut_list -> is_uni_eigen_loop_indexing t cut_list
+    | None -> is_uni_eigen_loop_indexing t [Index.All] )
+  | (UVector | URowVector), [Index.Single (_ : 'a Expr.Fixed.t)] -> true
+  | UMatrix, [Single (_ : 'a Expr.Fixed.t); Single (_ : 'a Expr.Fixed.t)] ->
+      true
+  (* None of the below contain single cell access*)
+  | (UReal | UInt | UMathLibraryFunction | UFun (_, Void, _, _)), _ -> false
+  | (UVector | URowVector), _ -> false
+  | UMatrix, _ -> false
+
+(**
+ * Specialization to find demotable (SoA -> AoS) objects
+ *  that carries logic on whether we are in a for loop or not.
+ * (TODO: Could be combined with other query demotable above.)
+ *)
+let rec query_demotable_exprs_loop in_loop Expr.Fixed.({pattern; _}) =
+  let query_expr = query_demotable_exprs_loop in_loop in
+  match pattern with
+  | FunApp (kind, (exprs : Expr.Typed.Meta.t Expr.Fixed.t list)) ->
+      query_demotable_funkinds_loop query_demotable_exprs_loop in_loop kind
+        exprs
+  | Var (_ : string) | Lit ((_ : Expr.Fixed.Pattern.litType), (_ : string)) ->
+      Set.Poly.empty
+  | TernaryIf (predicate, texpr, fexpr) ->
+      Set.Poly.union
+        (Set.Poly.union (query_expr predicate) (query_expr texpr))
+        (query_expr fexpr)
+  | Indexed (Expr.Fixed.({pattern= Var name; meta= {type_; _}}), indexed) ->
+      let query_index = map_index query_expr in
+      if in_loop && is_uni_eigen_loop_indexing type_ indexed then
```

Comment:
It doesn't matter if the single indexing uses the loop variable? It's just any single indexing inside any loop?

Reply:
I guess the loop thing is more of a rule of thumb because we don't want to single index very much?
We'll want to keep an eye on how this interacts with loop unrolling

Reply:
Yeah so the scheme here is that with the normal `Eigen::Matrix<var, -1, -1>` (aka Array of Structs) we are pretty much just working with a simple struct like 

```cpp
struct vari {
  double value_;
  double adjoint_;
};

struct matrix {
  vari*[] data_; // array of structs
  int rows_;
  int cols_;
};
```

So we have an array of pointers (the `vari**`) where each element of the array points to a struct that holds two doubles. Doing single cell access and assignment is pretty easy as we are just replacing one pointer in the array with another pointer. Replacing any pointer in that `vari**` still keeps the old pointer's actual memory on the stack so previous operations can use the value's the previous pointer pointed to. But with the new Struct of Arrays our structure looks something like 

```cpp
vari {
  // these are arrays now
  double[] value_;
  double[] adjoint_;
};

struct matrix {
 vari* data_; // struct of arrays
 int rows_;
 int cols_;
};
```

So now instead of an array to pointers of structs (`vari**`) we just have one pointer which references an object holding two arrays of doubles. This is very good for optimizations, but very bad when we need to do single cell access and assignment. With a struct of arrays, when we do single cell access we need to index into both of the sub arrays and create a new `vari` that represents that particular "cell". And when we do assignment it's even worse because we can't just swap out a pointer. We have to assign to the positions in the underlying array, but then put a callback on the callback stack to revert that value as we travel backwards in the reverse pass. 

Overall it's about 3-4x as slow to do single cell assignments and subsetting with structs of arrays, but conversely it can be easily 3-4x+ faster to use structs of arrays when we do vectorized operations. It's fine to do a few things by single cells as long as the dominant cost of the model is the actual math computations. So our heuristic to disallow the new matrix type in loops is because that type of code is normally going to need lots of single cell accessing/assignment. But idt users are normally modifying large amounts of individual cells of matrices and vectors without being in a loop.